ext/standard/uuid.c

```diff
@@ -0,0 +1,662 @@
+  +----------------------------------------------------------------------+
+  | Copyright (c) 1997-2017 The PHP Group                                |
+  +----------------------------------------------------------------------+
+  | This source file is subject to version 3.01 of the PHP license,      |
+  | that is bundled with this package in the file LICENSE, and is        |
+  | available through the world-wide-web at the following url:           |
+  | http://www.php.net/license/3_01.txt                                  |
+  | If you did not receive a copy of the PHP license and are unable to   |
+  | obtain it through the world-wide-web, please send a note to          |
+  | license@php.net so we can mail you a copy immediately.               |
+  +----------------------------------------------------------------------+
+  | Author: Richard Fussenegger <php@fleshgrinder.com>                   |
+  +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include "php_uuid.h"
+
+#include "zend_exceptions.h"
+#include "ext/spl/spl_exceptions.h"
+#include "md5.h"
+#include "php_random.h"
+#include "sha1.h"
+
+PHPAPI zend_class_entry *php_ce_UUID;
+PHPAPI zend_class_entry *php_ce_UUIDParseException;
+
+static const uint8_t UUID_VERSION_MIN = 0;
+static const uint8_t UUID_VERSION_MAX = 15;
+
+static const uint8_t UUID_HEX_LEN = sizeof(php_uuid_hex) - 1;
+static const uint8_t UUID_STRING_LEN = sizeof(php_uuid_string) - 1;
+
+static const char UUID_BYTES_PROP[] = "bytes";
+static const uint8_t UUID_BYTES_PROP_LEN = sizeof(UUID_BYTES_PROP) - 1;
+
+static const char UUID_EX_INPUT_PROP[] = "input";
+static const uint8_t UUID_EX_INPUT_PROP_LEN = sizeof(UUID_EX_INPUT_PROP) - 1;
+
+static const char UUID_EX_POSITION_PROP[] = "position";
+static const uint8_t UUID_EX_POSITON_PROP_LEN = sizeof(UUID_EX_POSITION_PROP) - 1;
+
+static const char URN_PREFIX[] = "urn:uuid:";
+static const uint8_t URN_PREFIX_LEN = sizeof(URN_PREFIX) - 1;
+
+static zend_always_inline void set_variant_rfc4122(php_uuid *uuid)
+{
+	uuid->bytes[8] = (uuid->bytes[8] & 0x3F) | 0x80;
+}
+
+static zend_always_inline void php_uuid_set_version(php_uuid *uuid, const uint8_t version)
+{
+	assert(UUID_VERSION_MIN <= version && version <= UUID_VERSION_MAX);
+
+	uuid->bytes[6] = (uuid->bytes[6] & 0x0F) | (version << 4);
+}
+
+PHPAPI void php_uuid_create_v3(php_uuid *uuid, const php_uuid *namespace, const char *name, const size_t name_len)
+{
+	PHP_MD5_CTX context;
+	unsigned char digest[16];
+
+	PHP_MD5Init(&context);
+	PHP_MD5Update(&context, namespace->bytes, PHP_UUID_LEN);
+	PHP_MD5Update(&context, name, name_len);
+	PHP_MD5Final(digest, &context);
+	memcpy(uuid->bytes, digest, PHP_UUID_LEN);
+	set_variant_rfc4122(uuid);
+	php_uuid_set_version(uuid, PHP_UUID_VERSION_3_NAME_BASED_MD5);
+}
+
+PHPAPI int php_uuid_create_v4(php_uuid *uuid, const zend_bool throw)
+{
+	int result = php_random_bytes(uuid, PHP_UUID_LEN, throw);
+
+	if (result == SUCCESS) {
```

Comment:
Assuming throw=0 and result == FAILURE this is just going to return garbage with no way to detect, right?